contracts/Pool.sol

```diff
@@ -14,260 +15,368 @@ import "./IBorrowersRegistry.sol";
-        totalBorrowed = totalBorrowed + _amount;
+    emit Deposit(msg.sender, msg.value, block.timestamp);
+  }
 
-        payable(msg.sender).transfer(_amount);
+  /**
+   * Withdraws selected amount from the user deposits
+   * @dev _amount the amount to be withdrawn
+  **/
+  function withdraw(uint256 _amount) external nonReentrant {
+    _accumulateDepositInterests(msg.sender);
 
-        updateRates();
+    _burn(msg.sender, _amount);
 
-        emit Borrowing(msg.sender, _amount, block.timestamp);
-    }
+    payable(msg.sender).transfer(_amount);
 
-    /**
-     * Repays the message value
-     * It updates user borrowed balance, total borrowed amount and rates
-    **/
-    function repay() payable external {
-        accumulateBorrowingInterests(msg.sender);
+    _updateRates();
 
-        require(getBorrowed(msg.sender) >= msg.value, "You are trying to repay more that was borrowed.");
+    emit Withdrawal(msg.sender, _amount, block.timestamp);
+  }
 
-        borrowed[msg.sender] = borrowed[msg.sender] - msg.value;
-        totalBorrowed = totalBorrowed - msg.value;
 
-        updateRates();
+  /**
+   * Borrows the specified amount
+   * It updates user borrowed balance, total borrowed amount and rates
+   * @dev _amount the amount to be borrowed
+  **/
+  function borrow(uint256 _amount) payable external canBorrow {
+    require(address(this).balance >= _amount, "There is not enough funds in the pool to fund the loan.");
+    require(totalSupply() - totalBorrowed() >= _amount, "There is no enough deposit in the pool to fund the loan.");
 
-        emit Repayment(msg.sender, msg.value, block.timestamp);
-    }
+    _accumulateBorrowingInterests(msg.sender);
 
+    borrowed[msg.sender] += _amount;
+    borrowed[address(this)]+= _amount;
 
-    /* ========== VIEW FUNCTIONS ========== */
+    payable(msg.sender).transfer(_amount);
 
+    _updateRates();
 
-    /**
-     * Returns the current borrowed amount for the given user
-     * The value includes the interest rates owned at the current moment
-     * @dev _user the address of queried borrower
-    **/
-    function getBorrowed(address _user) public view returns(uint256) {
-        return borrowIndex.getIndexedValue(borrowed[_user], _user);
-    }
+    emit Borrowing(msg.sender, _amount, block.timestamp);
+  }
 
+  /**
+   * Repays the message value
+   * It updates user borrowed balance, total borrowed amount and rates
+  **/
+  function repay() payable external {
+    _accumulateBorrowingInterests(msg.sender);
 
-    /**
-     * Returns the current deposited amount for the given user
-     * The value includes the interest rates earned at the current moment
-     * @dev _user the address of queried depositor
-    **/
-    function getDeposits(address user) public view returns(uint256) {
-      return depositIndex.getIndexedValue(deposits[user], user);
-    }
+    require(borrowed[address(this)] >= msg.value, "You are trying to repay more that was borrowed by all users.");
```

Comment:
Is it even possible that the second `require()` statement would pass but this would fail?

Reply:
We could remove it.